server/src/main/java/org/elasticsearch/node/NodeConstructor.java

```diff
@@ -0,0 +1,1634 @@
+import org.elasticsearch.reservedstate.action.ReservedClusterSettingsAction;
+import org.elasticsearch.reservedstate.service.FileSettingsService;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.script.ScriptModule;
+import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.search.SearchModule;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.SearchUtils;
+import org.elasticsearch.search.aggregations.support.AggregationUsageService;
+import org.elasticsearch.shutdown.PluginShutdownService;
+import org.elasticsearch.snapshots.InternalSnapshotsInfoService;
+import org.elasticsearch.snapshots.RepositoryIntegrityHealthIndicatorService;
+import org.elasticsearch.snapshots.RestoreService;
+import org.elasticsearch.snapshots.SnapshotShardsService;
+import org.elasticsearch.snapshots.SnapshotsInfoService;
+import org.elasticsearch.snapshots.SnapshotsService;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.telemetry.TelemetryProvider;
+import org.elasticsearch.telemetry.tracing.Tracer;
+import org.elasticsearch.threadpool.ExecutorBuilder;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.Transport;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.upgrades.SystemIndexMigrationExecutor;
+import org.elasticsearch.usage.UsageService;
+import org.elasticsearch.watcher.ResourceWatcherService;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.common.util.CollectionUtils.concatLists;
+import static org.elasticsearch.core.Types.forciblyCast;
+
+/**
+ * Class uses to perform all the operations needed to construct a {@link Node} instance.
+ */
+class NodeConstructor {
+
+    /**
+     * Performs node construction operations
+     *
+     * @param initialEnvironment         the initial environment for this node, which will be added to by plugins
+     * @param serviceProvider            provides various service implementations that could be mocked
+     * @param forbidPrivateIndexSettings whether or not private index settings are forbidden when creating an index; this is used in the
+     *                                   test framework for tests that rely on being able to set private settings
+     */
+    static NodeConstructor construct(
```

Comment:
This terminology is already making my head hurt...a construct method on NodeConstructor which constructs a NodeConstructor, which is passed to the constructor of Node...

Could we use some different terminology, perhaps the common "builder" suffix? Could we also make this actually do the node construction (see my other comment in Node about the ctor args).

We can have the only public method be a static `build(Environment env)`. That could create an instance, which would then call the real Node ctor (which can be package private). Then have a MockNodeBuilder (would we even need a separate MockNode at the point since the services can be injected into the real builder?).

Reply:
I might be wrong here, but I would avoid using `builder` and `build` in this point: it does not seem to me this is really a builder, but more of a factory.
I would call this class `NodeFactory`, this method `create` and make it return a `Node`.

Inside, I would have used the `builder` terminology, as it is closer to a builder pattern - maybe breaking down as NodeFactoryBuilder (fully internal) -> NodeFactory -> Node (with package private ctor)

Reply:
I've tweaked the naming to `prepareConstruction` - doing everything needed before you actually create a `Node`